File: Website/Backend/app/schemas/asset.py
"""
Pydantic response schemas for the Asset API.

Naming convention boundary (SYSTEM_DESIGN.md §5.2):
  DB / Python: snake_case  (monthly_rent, sale_price, …)
  JSON output: camelCase   (monthlyRent, salePrice, …)

Resolution: alias_generator=to_camel + serialize_by_alias=True.
Every field is named in Python snake_case; the serialised JSON uses camelCase
automatically. No field-level aliases needed — the generator handles them all.

The frontend Asset type in src/types/asset.ts is the authority.
This schema must match it exactly; never change the frontend type to match here.
"""

from typing import Optional

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel


class AssetCoordinates(BaseModel):
    """{ lat, lng } — both names are already single-word so to_camel leaves them unchanged."""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        serialize_by_alias=True,
    )

    lat: float
    lng: float


class AssetResponse(BaseModel):
    """
    Serialised representation of an Asset for GET /assets and GET /assets/{id}.

    Field names use Python snake_case; JSON output is camelCase via alias_generator.
    Matches frontend Asset interface in src/types/asset.ts exactly.

    PropertyType values:
      "retail_unit" | "ground_floor_commercial" | "office_unit" |
      "kiosk" | "warehouse" | "mixed_use" | "other"

    AvailabilityStatus values:
      "available" | "reserved" | "occupied" | "sold"

    ListingSource values:
      "seeded" | "public_registry" | "partner" | "direct" | "other"
    """

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        serialize_by_alias=True,
    )

    # Identity & location — always present
    id: str
    coordinates: AssetCoordinates
    address: str
    area: str
    size: Optional[float] = None

    # Commercial terms — either or both may be absent
    monthly_rent: Optional[float] = None
    sale_price: Optional[float] = None
    currency: str

    # Classification — always present
    property_type: str
    availability_status: str

    # Provenance — always present
    listing_source: str

    # Intelligence readiness — always present (0–100)
    coverage_score: int

File: Website/Frontend/src/types/asset.ts
// ─────────────────────────────────────────────────────────────────────────────
// Asset — the central domain object (SYSTEM_DESIGN.md §2.1)
//
// An Asset is a real commercial opportunity (rent or purchase) seeded by the
// platform. It is the stable spine onto which all intelligence attaches.
// Users discover Assets; they do not create or manage them (single-sided product).
//
// Supersedes: Site (src/types/site.ts) as the central object.
// Site may persist as an internal/future concept but the product is asset-first.
// ─────────────────────────────────────────────────────────────────────────────

export type PropertyType =
  | "retail_unit"
  | "ground_floor_commercial"
  | "office_unit"
  | "kiosk"
  | "warehouse"
  | "mixed_use"
  | "other";

export type AvailabilityStatus =
  | "available"
  | "reserved"
  | "occupied"
  | "sold";

// Near-term: seeded/public. Long-term: partners, providers, direct onboarding.
export type ListingSource =
  | "seeded"
  | "public_registry"
  | "partner"
  | "direct"
  | "other";

export interface AssetCoordinates {
  lat: number;
  lng: number;
}

export interface Asset {
  // ── Identity & location ─────────────────────────────────────────────────────
  id: string;
  coordinates: AssetCoordinates;
  address: string;
  area: string;          // district / neighbourhood label (e.g. "Maadi", "Heliopolis")
  size?: number;         // sq metres; optional

  // ── Commercial terms ────────────────────────────────────────────────────────
  // The listed price the valuation service reasons against.
  // Either or both may be present depending on the listing type.
  monthlyRent?: number;  // asking monthly rent in local currency
  salePrice?: number;    // asking purchase price
  currency: string;      // ISO 4217, e.g. "EGP"

  // ── Classification ─────────────────────────────────────────────────────────
  propertyType: PropertyType;
  availabilityStatus: AvailabilityStatus;

  // ── Provenance ─────────────────────────────────────────────────────────────
  listingSource: ListingSource;

  // ── Intelligence readiness ─────────────────────────────────────────────────
  // Surfaced on asset selection, before concept choice (SYSTEM_DESIGN.md §2.4).
  // A coverage score of 0 means the platform has almost no intelligence about
  // this asset; 100 means all six data sources are present.
  coverageScore: number; // 0–100
}
